quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -571,46 +571,96 @@ impl Metastore for PostgresqlMetastore {
-    async fn stage_split(
+    #[instrument(skip(self, split_metadata_list))]
+    async fn stage_splits(
         &self,
         index_id: &str,
-        split_metadata: SplitMetadata,
+        split_metadata_list: Vec<SplitMetadata>,
     ) -> MetastoreResult<()> {
-        let split_metadata_json = serde_json::to_string(&split_metadata).map_err(|error| {
-            MetastoreError::JsonSerializeError {
-                struct_name: "SplitMetadata".to_string(),
-                message: error.to_string(),
+        let mut split_ids = Vec::with_capacity(split_metadata_list.len());
+        let mut time_range_start_list = Vec::with_capacity(split_metadata_list.len());
+        let mut time_range_end_list = Vec::with_capacity(split_metadata_list.len());
+        let mut tags_list = Vec::with_capacity(split_metadata_list.len());
+        let mut split_metadata_json_list = Vec::with_capacity(split_metadata_list.len());
+        let mut delete_opstamps = Vec::with_capacity(split_metadata_list.len());
+
+        for split_metadata in split_metadata_list {
+            let split_metadata_json = serde_json::to_string(&split_metadata).map_err(|error| {
+                MetastoreError::JsonSerializeError {
+                    struct_name: "SplitMetadata".to_string(),
+                    message: error.to_string(),
+                }
+            })?;
+            let time_range_start = split_metadata
+                .time_range
+                .as_ref()
+                .map(|range| *range.start());
+            let time_range_end = split_metadata.time_range.map(|range| *range.end());
+            let tags: Vec<String> = split_metadata.tags.into_iter().collect();
+
+            split_ids.push(split_metadata.split_id);
+            time_range_start_list.push(time_range_start);
+            time_range_end_list.push(time_range_end);
+            tags_list.push(sqlx::types::Json(tags));
+            split_metadata_json_list.push(split_metadata_json);
+            delete_opstamps.push(split_metadata.delete_opstamp as i64);
+        }
+
+        run_with_tx!(self.connection_pool, tx, {
+            let upserted_split_ids: Vec<String> = sqlx::query_scalar(r#"
+                INSERT INTO splits
+                    (split_id, time_range_start, time_range_end, tags, split_metadata_json, delete_opstamp, split_state, index_id)
+                SELECT
+                    split_id,
+                    time_range_start,
+                    time_range_end,
+                    ARRAY(SELECT json_array_elements_text(tags_json::json)) as tags,
+                    split_metadata_json,
+                    delete_opstamp,
+                    $7 as split_state,
+                    $8 as index_id
+                FROM
+                    unnest($1, $2, $3, $4, $5, $6)
+                    as tr(split_id, time_range_start, time_range_end, tags_json, split_metadata_json, delete_opstamp)
```

Comment:
ah I see what you meant in the comments about sqlx not supporting batch insert now T_T

Reply:
The bigger issue was dealing with the fact that SQLx only supports 1 dimensional arrays 😅